source/val/validate_cfg.cpp

```diff
@@ -610,6 +611,122 @@ spv_result_t StructuredControlFlowChecks(
+      if (label_inst->uses().size() != 1) {
+        return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+               << "For WebGPU, unreachable merge-blocks must only be "
+                  "referenced by a single merge instruction, and cannot be the "
+                  "target of an OpBranch.";
+      }
+    } else if (block->is_type(kBlockTypeContinue)) {
+      // 1. Find referencing loop and confirm that it is reachable.
+      BasicBlock* continue_header = function->GetContinueHeader(block);
+      assert(continue_header != nullptr);
+      if (!continue_header->reachable()) {
+        return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+               << "For WebGPU, unreachable continue-targe must be referenced "
+                  "by a reachable loop instruction.";
+      }
+
+      // 2. Confirm that continue-target has only a single back edge to loop
+      // block.
+      auto target_id = block->id();
+      auto header_id = continue_header->id();
+      bool seen_back_edge = false;
+      for (auto& edge : back_edges) {
+        auto from = edge.first;
+        if (from != target_id) continue;
+        if (seen_back_edge) {
+          return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+                 << "For WebGPU, unreachable continue-target must only have a "
+                    "single back edge.";
+        }
+        seen_back_edge = true;
+        auto to = edge.second;
+        if (to != header_id) {
+          return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+                 << "For WebGPU, unreachable continue-target must only have a "
+                    "back edge to the loop instruction.";
+        }
+      }
+
+      if (!seen_back_edge) {
+        return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+               << "For WebGPU, unreachable continue-target must have a "
+                  "back edge to the loop instruction.";
+      }
+
+      // 3. Check that the only instructions are OpLabel and OpBranch.
+      auto* label_inst = block->label();
+      auto* terminator_inst = block->terminator();
+      assert(label_inst != nullptr);
+      assert(terminator_inst != nullptr);
+
+      if (terminator_inst->opcode() != SpvOpBranch) {
+        return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+               << "For WebGPU, unreachable continue-target must terminate with "
+                  "OpBranch.";
+      }
+
+      auto label_idx = label_inst - &_.ordered_instructions()[0];
+      auto terminator_idx = terminator_inst - &_.ordered_instructions()[0];
+      if (label_idx + 1 != terminator_idx) {
+        return _.diag(SPV_ERROR_INVALID_CFG, _.FindDef(block->id()))
+               << "For WebGPU, unreachable continue-target must only contain "
+                  "an OpLabel and an OpBranch instruction.";
+      }
+
+      // 4. Use OpLabel instruction to confirm there is only 1 use of the block.
+      if (label_inst->uses().size() != 1) {
```

Comment:
What about?
```
%loop = OpLabel
OpLoopMerge %merge %continue None
OpBranch %body
%body = OpLabel
OpSelectionMerge %continue
OpBranchConditional %c %then %else
%then = OpLabel
OpReturn
%else = OpLabel
OpReturn
%continue = OpLabel
OpBranch %loop
%merge = OpLabel
OpUnreachable
```

`%continue` is referenced twice, but this strikes me as valid. Maybe this and the merge check should be that all uses must be merge instructions?

Reply:
For you specific example is isn't valid, because %continue is being used a continue-target for OpLoopMerge and a merge-block for OpSelectionMerge, and those have mutually incompatible rules for termination of the block. That being said, there might an issue with if a block is referenced from multiple of the same type of header.

For selection merge, it seems likely that someone might want all of the degenerate merge-blocks collesced into a single example, so that defiantely needs to be investigated.

For loop merge, given that the target has to have a branch back to the header, I don't think it will be possible to encode multiple usages without breaking that rule.


... never mind, there is another rule in the validator that requires uniqueness of merge-blocks... I added related tests and refactored the code to more closely match what the spec says.

Reply:
This strikes me as bug in the environment spec. @dneto0 what do you think. My example could be generated by a frontend easily enough. To make it legal would require introducing or changing the selection's merge target.

Reply:
Right, merge blocks must be unique (1-1 mapping back to header).

Yes, there is nothing stopping a continue target from also being a merge block.  That's @alan-baker's example just above here.

Another troublesome case is a continue target that is named by two different loop-merge instructions.  something like this:

```
%cond0 = ...
%cond1 = ...

%outer = OpLabel
OpLoopMerge %outermerge %continue None
OpBranch %inner

        %inner = OpLabel
        OpLoopMerge %innermerge %continue
        OpBranchConditional %cond0 %innermerge %continue

                %continue = OpLabel
                OpBranchConditional %cond1 %inner %outer ; branches to both loop headers

        %innermerge = OpLabel
        OpBranch %outermerge

%outermerge = OpLabel
OpReturn
```

Reply:
I agree a block can be both a continue-target and a merge-block per @alan-baker's example. I talked with @dneto0 about the multiple loop header case. For the case of an unreachable continue-target, the rule about ending in an OpBranch to the header means that having multiple loops using the same target will fail either due to not ending in an OpBranch, or not being able to fulfill the rule about branching back to the referencing header for atleast one of them. Though the code still needs to properly handle multiple referencing headers.

Reply:
I have added in some test cases for the block being both a merge and continue.